src/charm.py

```diff
@@ -2,134 +2,108 @@
 
-    allowed_headers = [
-        "cache-control",
-        "content-transfer-encoding",
-        "content-type",
-        "grpc-timeout",
-        "keep-alive",
-        "user-agent",
-        "x-accept-content-transfer-encoding",
-        "x-accept-response-streaming",
-        "x-grpc-web",
-        "x-user-agent",
-        "custom-header-1",
-    ]
+        self.charm_reconciler = CharmReconciler(self)
 
-    virtual_host = api.route.VirtualHost(
-        name="local_service",
-        domains=["*"],
-        routes=[
-            api.route.Route(
-                match=api.route.RouteMatch(prefix="/"),
-                route=api.route.RouteAction(
-                    cluster=cluster,
-                    max_grpc_timeout=timedelta(seconds=60),
-                ),
+        self.leadership_gate = self.charm_reconciler.add(
+            component=LeadershipGateComponent(
+                charm=self,
+                name="leadership-gate",
             )
-        ],
-        cors=api.route.CorsPolicy(
-            allow_origin=["*"],
-            allow_methods=",".join(allowed_methods),
-            allow_headers=",".join(allowed_headers),
-            max_age="1728000",
-            expose_headers="grpc-status,grpc-message,custom-header-1",
-        ),
-    )
+        )
 
-    filter_listener = api.listener.Filter(
-        name="envoy.http_connection_manager",
-        config=hcm.HttpConnectionManager(
-            codec_type=hcm.HttpConnectionManagerCodecType.AUTO,
-            stat_prefix="ingress_http",
-            route_config=api.RouteConfiguration(
-                name="local_route",
-                virtual_hosts=[virtual_host],
+        # TODO: Change this from SDI to the service-info lib
+        self.grpc = self.charm_reconciler.add(
+            component=SdiRelationDataReceiverComponent(
+                charm=self,
+                name="relation:grpc",
+                relation_name="grpc",
             ),
-            http_filters=[
-                hcm.HttpFilter(name="envoy.grpc_web"),
-                hcm.HttpFilter(name="envoy.cors"),
-                hcm.HttpFilter(name="envoy.router"),
-            ],
-        ),
-    )
-    return api.Listener(
-        name="listener_0",
-        address=api.core.Address(
-            socket_address=api.core.SocketAddress(
-                address="0.0.0.0",
-                port_value=port,
-            )
-        ),
-        filter_chains=[api.listener.FilterChain(filters=[filter_listener])],
-    )
-
-
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
+            depends_on=[self.leadership_gate],
+        )
 
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
+        # Should this Component block the charm if it is not available?
+        # It is a requirement of the Kubeflow bundle, not the envoy charm.
+        # Envoy doesn't  NEED this relation, but Charmed Kubeflow using this for
```

Comment:
Yeah I'm torn on this one for the same reasons you are.  On one hand, you're right that we're never going to actually use this without the ingress.  But then we could say the same for all our dashboard charms and none of them have this requirement.

I lean toward being consistent (and not *requiring* an ingress in the charm), but I don't like that we can easily break this at bundle level

Reply:
We don't have the ingress requirement in other charms, but I don't think that is by design. [`katib-ui`](https://github.com/canonical/katib-operators/blob/main/charms/katib-ui/src/charm.py#L173) for instance is definitely using the ingress relation on every single event. I think SDI does something about missing relations (errors out?) it shows in the status message.

To me this component, alongside metadata-grpc-service (mlmd), is meant to be used by the Kubeflow Pipelines, which is also tightly coupled with other components, like the Kubeflow dashboard and the Istio ingress. That being said, and because we haven't defined a strategy for standalone components, I think we can be fine w/o requiring the ingress relation. I'd suggest to create an issue to start the discussion and log somewhere in this charm that the ingress relation is not set, if used in a Kubeflow + Istio environment, it has to be set, it can be a WARNING.

Reply:
Yeah I like discussing this as a general strategy.  There's definitely a testing benefit to letting the charm run when ingress is omitted, but then that same "feature" opens our real product up to a potential bug...  Added an issue as canonical/bundle-kubeflow#852

Reply:
As for giving a warning if the relation is missing, I'll have to think on how to do that.

Reply:
This warning is added now

Reply:
Sounds good.